File: site/05-Heaps-and-Priority-Queues/js/min_heap.js
class MinHeap {
    constructor() {
        this.heap = [];
    }

    getParentIndex(i) { return Math.floor((i - 1) / 2); }
    getLeftChildIndex(i) { return 2 * i + 1; }
    getRightChildIndex(i) { return 2 * i + 2; }

    insert(val) {
        this.heap.push(val);
        this.heapifyUp(this.heap.length - 1);
    }

    extractMin() {
        if (this.heap.length === 0) return null;
        if (this.heap.length === 1) return this.heap.pop();
        
        const minVal = this.heap[0];
        this.heap[0] = this.heap.pop();
        this.heapifyDown(0);
        return minVal;
    }

    heapifyUp(index) {
        const parent = this.getParentIndex(index);
        if (index > 0 && this.heap[index] < this.heap[parent]) {
            [this.heap[index], this.heap[parent]] = [this.heap[parent], this.heap[index]];
            this.heapifyUp(parent);
        }
    }

    heapifyDown(index) {
        const left = this.getLeftChildIndex(index);
        const right = this.getRightChildIndex(index);
        let smallest = index;

        if (left < this.heap.length && this.heap[left] < this.heap[smallest]) {
            smallest = left;
        }
        if (right < this.heap.length && this.heap[right] < this.heap[smallest]) {
            smallest = right;
        }
        if (smallest !== index) {
            [this.heap[index], this.heap[smallest]] = [this.heap[smallest], this.heap[index]];
            this.heapifyDown(smallest);
        }
    }

    isEmpty() {
        return this.heap.length === 0;
    }
}

// Test Heap
const heap = new MinHeap();
console.log("Inserting 3, 1, 6, 5, 2, 4 into MinHeap:");
[3, 1, 6, 5, 2, 4].forEach(v => heap.insert(v));

const result = [];
while (!heap.isEmpty()) {
    result.push(heap.extractMin());
}
console.log("Extracted order:", result.join(" "));

File: docs/05-Heaps-and-Priority-Queues/python/min_heap.py
class MinHeap:
    def __init__(self):
        self.heap = []

    def get_parent_index(self, index):
        return (index - 1) // 2

    def get_left_child_index(self, index):
        return 2 * index + 1

    def get_right_child_index(self, index):
        return 2 * index + 2

    def insert(self, val):
        self.heap.append(val)
        self.heapify_up(len(self.heap) - 1)

    def extract_min(self):
        if not self.heap:
            return None
        if len(self.heap) == 1:
            return self.heap.pop()
        
        min_val = self.heap[0]
        self.heap[0] = self.heap.pop()
        self.heapify_down(0)
        return min_val

    def peek(self):
        return self.heap[0] if self.heap else None

    def heapify_up(self, index):
        parent = self.get_parent_index(index)
        if index > 0 and self.heap[index] < self.heap[parent]:
            self.heap[index], self.heap[parent] = self.heap[parent], self.heap[index]
            self.heapify_up(parent)

    def heapify_down(self, index):
        left = self.get_left_child_index(index)
        right = self.get_right_child_index(index)
        smallest = index
        
        if left < len(self.heap) and self.heap[left] < self.heap[smallest]:
            smallest = left
        if right < len(self.heap) and self.heap[right] < self.heap[smallest]:
            smallest = right
            
        if smallest != index:
            self.heap[index], self.heap[smallest] = self.heap[smallest], self.heap[index]
            self.heapify_down(smallest)

if __name__ == "__main__":
    heap = MinHeap()
    print("Inserting 3, 1, 6, 5, 2, 4 into MinHeap:")
    for v in [3, 1, 6, 5, 2, 4]:
        heap.insert(v)
        
    print("Heap contents peek:", heap.peek())
    print("Extracting minimum elements in order:")
    while len(heap.heap) > 0:
        print(heap.extract_min(), end=" ")
    print()
